src/Common/src/CoreWCF/Runtime/TaskHelpers.cs

```diff
@@ -29,120 +30,124 @@ public static async Task AsyncWait<TException>(this Task task)
-        public static Task<TResult> ToApm<TResult>(this Task<TResult> task, AsyncCallback callback, object state)
+        public static IAsyncResult ToApm<T>(this Task<T> task, AsyncCallback callback, object state)
+            => ToApm<T>(new ValueTask<T>(task), callback, state);
+
+        /// <summary>
+        /// Helper method to convert from Task async method to "APM" (IAsyncResult with Begin/End calls)
+        /// </summary>
+        public static IAsyncResult ToApm<T>(this ValueTask<T> valueTask, AsyncCallback callback, object state)
         {
-            // When using APM, the returned IAsyncResult must have the passed in state object stored in AsyncState. This
-            // is so the callback can regain state. If the incoming task already holds the state object, there's no need
-            // to create a TaskCompletionSource to ensure the returned (IAsyncResult)Task has the right state object.
-            // This is a performance optimization for this special case.
-            if (task.AsyncState == state)
+            var result = new AsyncResult<T>(valueTask, callback, state);
+            if (result.CompletedSynchronously)
             {
-                if (callback != null)
-                {
-                    task.ContinueWith((antecedent, obj) =>
-                    {
-                        var callbackObj = obj as AsyncCallback;
-                        callbackObj(antecedent);
-                    }, callback, CancellationToken.None, TaskContinuationOptions.HideScheduler, TaskScheduler.Default);
-                }
-                return task;
+                result.ExecuteCallback();
             }
-
-            // Need to create a TaskCompletionSource so that the returned Task object has the correct AsyncState value.
-            var tcs = new TaskCompletionSource<TResult>(state);
-            var continuationState = Tuple.Create(tcs, callback);
-            task.ContinueWith((antecedent, obj) =>
+            else if (callback != null)
             {
-                var tuple = obj as Tuple<TaskCompletionSource<TResult>, AsyncCallback>;
-                var tcsObj = tuple.Item1;
-                var callbackObj = tuple.Item2;
-                if (antecedent.IsFaulted) tcsObj.TrySetException(antecedent.Exception.InnerException);
-                else if (antecedent.IsCanceled) tcsObj.TrySetCanceled();
-                else tcsObj.TrySetResult(antecedent.Result);
-
-                if (callbackObj != null)
+                valueTask.AsTask().ContinueWith((_, asyncResult) =>
                 {
-                    callbackObj(tcsObj.Task);
-                }
-            }, continuationState, CancellationToken.None, TaskContinuationOptions.HideScheduler, TaskScheduler.Default);
-            return tcs.Task;
+                    ((AsyncResult<T>)asyncResult).ExecuteCallback();
+                }, result, CancellationToken.None, TaskContinuationOptions.RunContinuationsAsynchronously, TaskScheduler.Default);
```

Comment:
I made one new change by: 

1. adding `RunContinuationsAsynchronously` inspired by the .Net framework  code so that continuations are not allowed to executed sync on the thread where the task is completed. 
This should work fine, but has not been stress tested by me (tests still pass).

I am happy to revert this change if you think the old way was better

2. removing `HideScheduler` (which should not have any effect since TaskScheduler.Default will be the ambient scheduler both with and without the flag)